File: LeetCode Problems/Other/Medium/38. Count and Say/solution.py
class Solution:
    def encode(self, en: str) -> str:
        count = 1
        new_en = ''

        for i in range(len(en)):
            # when the characters aren't matching OR it is the last character of the sequence
            if i == len(en) - 1 or en[i] != en[i + 1]:
                new_en += str(count) + en[i]
                count = 1
            else:
                count += 1
        
        return new_en

    def countAndSay(self, n: int) -> str:
        dp = ['']*n
        dp[0] = '1'

        for i in range(1, n):
            dp[i] = self.encode(dp[i - 1])

        return dp[n - 1]

File: LeetCode Problems/Other/Medium/38. Count and Say/solution.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    string encode(string en){
        int count = 1;
        string new_en = "";

        for(int i = 0; i < en.size(); i++){
            if(i == en.size() - 1 || en[i] != en[i + 1]){
                new_en += to_string(count);
                new_en += en[i];
                count = 1;
            }
            else{
                count++;
            }
        }

        return new_en;
    }

    string countAndSay(int n) {
        string prev, current;
        prev = '1';    

        for(int i = 1; i < n; i++){
            current = this->encode(prev);
            prev = current;
        }

        return prev;
    }
};